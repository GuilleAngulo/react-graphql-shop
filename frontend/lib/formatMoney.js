import { country } from '../config';

export default function(amount) {

  const options = [
    {
      locales: 'en-US',
      option: {
        style: 'currency',
        currency: 'USD',
        minimumFractionDigits: 2,
      },
    },
    {
      locales: 'pt-BR',
      option: {
        style: 'currency',
        currency: 'BRL',
        minimumFractionDigits: 2,
      },
    },
    {
      locales:  'es-ES',
      option: {
        style: 'currency',
        currency: 'EUR',
        minimumFractionDigits: 2,
      },
    },
  ];

  const { locales, option } = options.find(option => option.locales === country);

  // if its a whole, dollar amount, leave off the .00
  if (amount % 100 === 0) option.minimumFractionDigits = 0;

  const formatter = new Intl.NumberFormat(locales, option);
  return formatter.format(amount / 100);
};