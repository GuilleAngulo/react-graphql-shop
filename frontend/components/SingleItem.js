import React, { Component } from 'react';
import gql from 'graphql-tag';
import { Query } from 'react-apollo';
import Error from './ErrorMessage';
import styled from 'styled-components';
import Head from 'next/head';

import formatMoney from '../lib/formatMoney';
import AddToCart from './AddToCart';
import { appTitle } from '../config';

const SingleItemStyles = styled.div`
    max-width: 1200px;
    margin: 2rem auto;
    box-shadow: ${props => props.theme.bs};
    display: grid;
    grid-auto-columns: 1fr;
    grid-auto-flow: column;
    min-height: 800px;
    img {
        width: 100%;
        height: 100%;
        object-fit: contain;
        margin: 10px;
    }
    .details {
        margin: 5rem;
        font-size: 2rem;
    }
    p.title {
        background: ${props => props.theme.darkOrange};
        display: inline;
        line-height: 1.3;
        font-size: 4rem;
        text-align: center;
        color: white;
        padding: 0 1rem;
    }
    p.price {
        font-size: 16px;
    }
    .button {
    width: 100%;
    & > * {
      background: ${props => props.theme.lightGrey};
      border: 0;
      font-size: 2rem;
      padding: 1rem;
      cursor: pointer;
      border-radius: 10px;
    }
  }
`;

const SINGLE_ITEM_QUERY = gql`
    query SINGLE_ITEM_QUERY($id: ID!) {
        item(where: { id: $id }) {
            id
            title
            description
            image
            largeImage
            price
        }
    }
`;

class SingleItem extends Component {
    render() {
        return (
            <Query 
            query={SINGLE_ITEM_QUERY}
            variables={{
                id: this.props.id,
            }}
            >
                {({ error, loading, data }) => {
                    if (error) return <Error error={error} />;
                    if (loading) return <p>Loading...</p>;
                    if (!data.item) return <p>No item found for {this.props.id}</p>
                    const item = data.item;
                    return (
                    <SingleItemStyles>
                        <Head>
                            <title> {appTitle} | {item.title} </title>
                        </Head>
                        <img src={item.largeImage} alt={item.title}/> 
                        <div className="details">
                            <h2 className="title">{item.title}</h2>
                            <p>{item.description}</p>
                            <p className="price">{formatMoney(item.price)}</p>
                            <div className="button">
                                <AddToCart className="add" item={ item }/>
                            </div>
                        </div>
                    </SingleItemStyles>
                    );
                }}
            </Query>
        );
    }
}

export default SingleItem;
export { SINGLE_ITEM_QUERY };