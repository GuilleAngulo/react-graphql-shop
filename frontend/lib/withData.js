import withApollo from 'next-with-apollo';
import ApolloClient from 'apollo-boost';
import { LOCAL_STATE_QUERY } from '../components/Cart';

import { endpoint } from '../config';

function createClient({ headers }) {
  return new ApolloClient({
    uri: process.env.NODE_ENV === 'development' ? endpoint : endpoint,
    request: operation => {
      operation.setContext({
        fetchOptions: {
          credentials: 'include',
        },
        headers,
      });
    },
    //LOCAL DATA
    clientState: {
      resolvers: {
        Mutation: {
          toogleCart(_, variables, { cache }) {
            //Read the cartOpen value from the cache
            const { cartOpen } = cache.readQuery({
              query: LOCAL_STATE_QUERY,
            });
            //Write the cart State to the opposite
            const data = {
              data: { cartOpen: !cartOpen }
            };
            cache.writeData(data);
            return data;
          },
        },
      },
      defaults: {
        cartOpen: false,
      },
    },
  });
}

export default withApollo(createClient);
