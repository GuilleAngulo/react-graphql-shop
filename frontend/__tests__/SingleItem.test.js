import { mount } from 'enzyme';
import toJSON from 'enzyme-to-json';
import wait from 'waait';
import SingleItem, { SINGLE_ITEM_QUERY } from '../components/SingleItem';
import { MockedProvider } from 'react-apollo/test-utils';
import { fakeItem } from '../lib/testUtils';

describe('<SingleItem />', () => {
    it('renders with proper data', async () => {
        const mocks = [
            {
                //When someone makes a request with this qwuery and variable combo
                request: { 
                    query: SINGLE_ITEM_QUERY,
                    variables: { id: '123'},
                },
                //Return this fake data (mocked data)
                result: {
                    data: { item: fakeItem(),},
                },
            },
        ];
        const wrapper = mount(
        <MockedProvider mocks={mocks}>
            <SingleItem id="123" />
        </MockedProvider>
        );

        //console.log(wrapper.debug());
        expect(wrapper.text()).toContain('Loading...');
        await wait();
        wrapper.update();
        
        
        expect(toJSON(wrapper.find('h2'))).toMatchSnapshot();

        expect(toJSON(wrapper.find('img'))).toMatchSnapshot();

        expect(toJSON(wrapper.find('p'))).toMatchSnapshot();
    });


    it('errors with a not found item', async () => {
        const mocks = [
            {
                request: { 
                    query: SINGLE_ITEM_QUERY,
                    variables: { id: '123'},
                },
                result: {
                    errors: [{
                        message: 'Items not found',
                    }], 
                },
            },
        ];

        const wrapper = mount(
            <MockedProvider mocks={mocks}>
                <SingleItem id="123" />
            </MockedProvider>
            );
        await wait();
        wrapper.update();
        //console.log(wrapper.debug());
        const item = wrapper.find('[data-test="graphql-error"]');
        //console.log(item.debug());
        expect(item.text()).toContain('Items not found');
        expect(toJSON(item)).toMatchSnapshot();

    });
});

