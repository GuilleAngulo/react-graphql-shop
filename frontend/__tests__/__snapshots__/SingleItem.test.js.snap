// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SingleItem /> errors with a not found item 1`] = `
<p
  data-test="graphql-error"
>
  <strong>
    Shoot!
  </strong>
  Items not found
</p>
`;

exports[`<SingleItem /> renders with proper data 1`] = `
<h2
  className="title"
>
  dogs are best
</h2>
`;

exports[`<SingleItem /> renders with proper data 2`] = `
<img
  alt="dogs are best"
  src="dog.jpg"
/>
`;

exports[`<SingleItem /> renders with proper data 3`] = `
Array [
  <p>
    dogs
  </p>,
  <p
    className="price"
  >
    $50
  </p>,
]
`;
